### PR TITLE
fix exception handling

### LINKED ISSUE #1009: PyPANDA Panda.hook2/_hook_mem exception handling
On 27eb888ce2235debe4e7032a84ccc15d4eebde8a, Panda.hook2 and Panda._hook_mem use these FFI callbacks:
```py
@hook_cb_type # Make CFFI know it's a callback. Different from _generated_callback for some reason?
def wrapper(*args, **kw):
    return fun(*args, **kw)
```
and
```py
@mem_hook_cb_type # Make CFFI know it's a callback. Different from _generated_callback for some reason?
def wrapper(*args, **kw):
    return fun(*args, **kw)
```

Other similar FFI callbacks have special handling for exceptions, for example in Panda.hook:
```py
@hook_cb_type # Make CFFI know it's a callback. Different from _generated_callback for some reason?
def wrapper(*args, **kw):
    try:
        r = fun(*args, **kw)
        #assert(isinstance(r, int)), "Invalid return type?"
        return r
    except Exception as e:
        # exceptions wont work in our thread. Therefore we print it here and then throw it after the
        # machine exits.
        self.hook_exit_exception = e
        self.end_analysis()
        # this works in all current callback cases. CFFI auto-converts to void, bool, int, and int32_t
        return 0
```

I wonder whether Panda.hook2 and Panda._hook_mem would also need special exception handling.
If this is the case, maybe these wrapper functions should be refactored into one separate function to avoid such inconsistencies in the future.

### PR DESCRIPTION
**Pull request by lacraig2**
- Creates a single exception value to exit at
- Checks exception value before entering new code
- standardizes `_run_and_catch` across the 6 code interfaces

FIXES #1009

### CODE REVIEW

File: panda/python/core/pandare/panda.py [hunk 1]
```diff
@@ -3088,8 +3101,7 @@ def decorator(fun):
 
             @hook_cb_type # Make CFFI know it's a callback. Different from _generated_callback for some reason?
```
Comment (AndrewFasano) [inline panda/python/core/pandare/panda.py]: You deleted this decorator above - but left it here. Not sure if that was intentional?

**Comment by AndrewFasano** [comment]
Thanks for cleaning up this mess, code looks good to me

File: panda/python/core/pandare/panda.py (hunk 1 above)
Comment (lacraig2) [reply panda/python/core/pandare/panda.py]: Good catch. Removed below.

**Comment by AndrewFasano** [comment]
Should we merge this?

**Comment by AndrewFasano** [comment]
I think I stomped on some of this code in a recent PR, I had thought this was already merged and fixed a bug in some of what this changes- sorry!

**Comment by lacraig2** [comment]
Superseded by #1149

**Pull request closed without merging**